src/main/java/org/spongepowered/common/entity/player/tab/SpongeTabListEntry.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.entity.player.tab;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.base.Objects;
+import net.minecraft.network.play.server.S38PacketPlayerListItem;
+import org.spongepowered.api.entity.living.player.gamemode.GameMode;
+import org.spongepowered.api.entity.living.player.tab.TabList;
+import org.spongepowered.api.entity.living.player.tab.TabListEntry;
+import org.spongepowered.api.profile.GameProfile;
+import org.spongepowered.api.text.Text;
+
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/*
+ * This is intentionally not a mixin of S38PacketPlayerListItem.AddPlayerData.
+ */
+public final class SpongeTabListEntry implements TabListEntry {
+
+    private SpongeTabList list;
+    private final GameProfile profile;
+    @Nullable private Text displayName;
+    private int latency;
+    private GameMode gameMode;
+    private boolean updateWithoutSend;
+
+    public SpongeTabListEntry(TabList list, GameProfile profile, @Nullable Text displayName, int latency, GameMode gameMode) {
+        checkState(list instanceof SpongeTabList, "list is not a SpongeTabList");
```

Comment:
Not sure if this is necessary, I mean otherwise it will simply throw a class cast exception which has even a kind of more helpful error message.